include/swift/Frontend/DiagnosticVerifier.h

```diff
@@ -18,30 +18,81 @@
 #ifndef SWIFT_FRONTEND_DIAGNOSTIC_VERIFIER_H
 #define SWIFT_FRONTEND_DIAGNOSTIC_VERIFIER_H
 
+#include "swift/AST/DiagnosticConsumer.h"
 #include "swift/Basic/LLVM.h"
 
 namespace swift {
-  class DependencyTracker;
-  class FileUnit;
-  class SourceManager;
-  class SourceFile;
-
-  /// Set up the specified source manager so that diagnostics are captured
-  /// instead of being printed.
-  void enableDiagnosticVerifier(SourceManager &SM);
-
-  /// Verify that captured diagnostics meet with the expectations of the source
-  /// files corresponding to the specified \p BufferIDs and tear down our
-  /// support for capturing and verifying diagnostics.
-  ///
-  /// This returns true if there are any mismatches found.
-  bool verifyDiagnostics(SourceManager &SM, ArrayRef<unsigned> BufferIDs,
-                         bool autoApplyFixes, bool ignoreUnknown);
-
-  bool verifyDependencies(SourceManager &SM, const DependencyTracker &DT,
-                          ArrayRef<FileUnit *> SFs);
-  bool verifyDependencies(SourceManager &SM, const DependencyTracker &DT,
-                          ArrayRef<SourceFile *> SFs);
+class DependencyTracker;
+class FileUnit;
+class SourceManager;
+class SourceFile;
+
+// MARK: - DependencyVerifier
+bool verifyDependencies(SourceManager &SM, const DependencyTracker &DT,
+                        ArrayRef<FileUnit *> SFs);
+bool verifyDependencies(SourceManager &SM, const DependencyTracker &DT,
+                        ArrayRef<SourceFile *> SFs);
+
+// MARK: - DiagnosticVerifier
+struct ExpectedFixIt;
+
+struct CapturedDiagnosticInfo {
+  llvm::SmallString<128> Message;
+  llvm::SmallString<32> FileName;
+  DiagnosticKind Classification;
+  SourceLoc Loc;
+  unsigned Line;
+  unsigned Column;
+  SmallVector<DiagnosticInfo::FixIt, 2> FixIts;
+};
+/// This class implements support for -verify mode in the compiler.  It
+/// buffers up diagnostics produced during compilation, then checks them
+/// against expected-error markers in the source file.
+class DiagnosticVerifier : public DiagnosticConsumer {
+  SourceManager &SM;
+  std::vector<CapturedDiagnosticInfo> CapturedDiagnostics;
+  ArrayRef<unsigned> BufferIDs;
+  bool AutoApplyFixes;
+  bool IgnoreUnknown;
+  bool UseColor;
```

Comment:
Let's leave color support out of this PR to try to keep this as close as possible to a pure refactoring.